cpp/include/ray/api/function_manager.h

```diff
@@ -0,0 +1,172 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <ray/api/serializer.h>
+#include "absl/utility/utility.h"
+
+#include <boost/callable_traits.hpp>
+#include <functional>
+#include <memory>
+#include <string>
+#include <tuple>
+#include <unordered_map>
+
+namespace ray {
+namespace api {
+
+namespace internal {
+
+template <class, class>
+struct AddType;
+
+template <class First, class... Second>
+struct AddType<First, std::tuple<Second...>> {
+  using type = std::tuple<First, Second...>;
+};
+
+template <class First, class Second>
+using AddType_t = typename AddType<First, Second>::type;
+
+enum ErrorCode {
+  OK = 0,
+  FAIL = 1,
+};
+
+struct VoidResponse {
+  int error_code;
+  std::string error_msg;
+
+  MSGPACK_DEFINE(error_code, error_msg);
+};
+
+template <typename T>
+struct Response {
+  int error_code;
+  std::string error_msg;
+  T data;
+
+  MSGPACK_DEFINE(error_code, error_msg, data);
+};
+
+template <typename T>
+inline static msgpack::sbuffer PackReturnValue(int error_code, std::string error_msg,
+                                               T result) {
+  return Serializer::Serialize(
+      Response<T>{error_code, std::move(error_msg), std::move(result)});
+}
+
+inline static msgpack::sbuffer PackReturnValue(int error_code,
+                                               std::string error_msg = "ok") {
+  return Serializer::Serialize(VoidResponse{error_code, std::move(error_msg)});
+}
+
+template <typename Function>
+struct Invoker {
+  static inline msgpack::sbuffer Apply(const Function &func, const char *data,
```

Comment:
Better to return a `std::shared_ptr<msgpack::sbuffer>`?

Reply:
msgpack::sbuffer supports move construct, i don't know what scene need shared_ptr now, if we need shared_ptr, we can update at that time.